master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala

```diff
@@ -666,23 +666,32 @@ private[celeborn] class Master(
     val shuffleKey = Utils.makeShuffleKey(requestSlots.applicationId, requestSlots.shuffleId)
 
     val availableWorkers = workersAvailable()
+    // reply false if all workers are unavailable
+    if (availableWorkers.isEmpty) {
+      logError(
+        s"Non available workers, offer slots for $numReducers reducers of $shuffleKey failed!")
+      context.reply(RequestSlotsResponse(StatusCode.SLOT_NOT_AVAILABLE, new WorkerResource()))
+      return
+    }
+
     val numAvailableWorkers = availableWorkers.size()
     val numWorkers = Math.min(
       Math.max(
         if (requestSlots.shouldReplicate) 2 else 1,
         if (requestSlots.maxWorkers <= 0) slotsAssignMaxWorkers
         else Math.min(slotsAssignMaxWorkers, requestSlots.maxWorkers)),
       numAvailableWorkers)
+
+    // We treated availableWorkers as a Circular Queue here.
     val startIndex = Random.nextInt(numAvailableWorkers)
+    val endIndex = startIndex + numWorkers
+
     val selectedWorkers = new util.ArrayList[WorkerInfo](numWorkers)
-    selectedWorkers.addAll(availableWorkers.subList(
-      startIndex,
-      Math.min(numAvailableWorkers, startIndex + numWorkers)))
-    if (startIndex + numWorkers > numAvailableWorkers) {
-      selectedWorkers.addAll(availableWorkers.subList(
-        0,
-        startIndex + numWorkers - numAvailableWorkers))
+    for (index <- startIndex until endIndex) {
+      val realIndex = index % numAvailableWorkers
+      selectedWorkers.add(availableWorkers.get(realIndex))
     }
+
```

Comment:
Wont this not be slower ? The range check is done for each `add`, and we are not able to leverage bulk add from sublist to array list as well.

The earlier code is definitely a bit dense - but that can be fixed by simply pulling out variables and making it cleaner ?

For example, something like (strawman proposal):
```
val selectedWorkers = new util.ArrayList[WorkerInfo](numWorkers)

def copySublist(from: Int): Unit = {
  val numEntriesToCopy = Math.min(numAvailableWorkers - from, numWorkers - selectedWorkers.size)
  selectedWorkers.addAll(availableWorkers.subList(from, from + numEntriesToCopy)
}

copySublist(startIndex)
if (selectedWorkers.size < numWorkers) {
  // wrap around to beginning to copy rest.
  copySublist(0)
}
```

Reply:
Hi, @mridulm, i know what you think, let's make a comparison. 

Randomly select 200 workers from 2000, here is the result, the difference is only on the millisecond level.

![image](https://github.com/apache/incubator-celeborn/assets/95013770/96b755ad-18ce-49ac-aa2d-d7878b3e398c)

Reply:
From a complexity point of view, this does not make sense - the input is a `LinkedList`, and so the `get` version is O(n^2) complexity, while the subList Over is (n).

Not sure how you benchmarked it - but for microbenchmarks, I would suggest using something like JMH: this specific example looks like it is overwhelmed by noise and precision issues - the sizes are too small , the number of iterations not high and the precision of measurement is ms.

Reply:
Even with the change to `ArrayList`, from `LinkedList`, it is still 1.8x faster (see [1]).

The readability is not that much different between both versions - particularly with better comments and less dense code. Not to mention, there is a clear perf difference.


[1]

Reply:
@mridulm, thanks for sharing your thought. 👍